src/python/pants/testutil/rule_runner.py

```diff
@@ -75,6 +75,53 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+class EngineErrorTrap:
+    """A context manager to catch `ExecutionError`s in tests and get the underlying exception.
+
+    Use like this:
+
+        with EngineErrorTrap() as trap:
+            rule_runner.request(OutputType, [input])
+        assert isinstance(trap.e, ValueError)
+
+    Will raise AssertionError if no ExecutionError occurred. Will re-raise all other exception
+    types.
+    """
+
+    _e: Exception | None = None
+
+    def __enter__(self) -> EngineErrorTrap:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is None:
+            raise AssertionError("Did not raise an exception! ExecutionError expected.")
+        if not issubclass(exc_type, ExecutionError):
+            # This is equivalent to re-raising the exception.
+            return None
+
+        if not len(exc_val.wrapped_exceptions) == 1:
+            formatted_errors = "\n\n".join(repr(e) for e in exc_val.wrapped_exceptions)
+            raise ValueError(
+                "Multiple underlying exceptions, but this helper function expected only one. "
+                "Use `with pytest.raises(ExecutionError)` directly and inspect "
+                "`exc.value.wrapped_exceptions`.\n\n"
+                f"Errors: {formatted_errors}"
+            )
+        self._e = exc_val.wrapped_exceptions[0]
+        return True
+
+    @property
+    def e(self) -> Exception:
+        if self._e is None:
+            raise AssertionError(
+                "`EngineErrorTrap.e` is not set. This happens when you try accessing the field "
+                "inside the context manager (i.e. the `with` block) and an ExecutionError has not "
+                "happened yet."
```

Comment:
This message is slightly misleading, since you would never reach the code accessing `trap.e` within the `with` block.

Reply:
I was surprised that it actually can happen! When the `ExecutionError` happens, it triggers `__exit__`. If the error is gracefully handled, the `with` block continues.

Reply:
So, to avoid surprises, perhaps it shouldn't... oh, right. hmm.. is there no way to exit the block prematurely, without throwing an exception to the outer scope?

Reply:
I don't think there's a way to do that, but I'm definitely not an expert on context managers. If you don't `return True` in `__exit__`, then the original exception will be thrown and blow everything up.